client/uploadable.go

```diff
@@ -166,33 +150,54 @@ func NewAutoUploadable(mqttClient MQTT.Client, edgeCfg *EdgeConfiguration, uploa
 
 	result.uploads = NewUploads()
 
-	result.client.Subscribe(func(requestID string, msg *protocol.Envelope) {
-		go result.messageHandler(requestID, msg)
-	})
-
 	return result, nil
 }
 
 // Connect AutoUploadable to the Ditto endpoint
-func (u *AutoUploadable) Connect() error {
+func (u *AutoUploadable) Connect(mqttClient MQTT.Client, edgeCfg *EdgeConfiguration) {
+	u.deviceID = edgeCfg.DeviceID
+	u.tenantID = edgeCfg.TenantID
+
+	config := ditto.NewConfiguration().
+		WithDisconnectTimeout(defaultDisconnectTimeout).
+		WithConnectHandler(
+			func(client *ditto.Client) {
+				u.connectHandler(client)
+			})
+
+	var err error
+	if u.client, err = ditto.NewClientMqtt(mqttClient, config); err != nil {
+		logger.Error(err)
+		return
+	}
+
+	u.client.Subscribe(func(requestID string, msg *protocol.Envelope) {
+		go u.messageHandler(requestID, msg)
+	})
+
+	if err := u.client.Connect(); err != nil {
+		logger.Error(err)
+	}
+
 	u.statusEvents.start(func(e interface{}) {
 		u.UpdateProperty(lastUploadProperty, e)
 	})
 
-	return u.client.Connect()
+	logger.Info("ditto client connected")
 }
 
 // Disconnect AutoUploadable from the Ditto endpoint and clean up used resources
 func (u *AutoUploadable) Disconnect() {
 	u.statusEvents.stop()
 
+	u.client.Disconnect()
 	u.client.Unsubscribe()
 
 	u.stopExecutor() //stop periodic triggers
 
 	u.uploads.Stop(time.Duration(u.cfg.StopTimeout)) // stop active uploads
 
-	u.client.Disconnect()
```

Comment:
Why this order is changed? Usually, the logic is:
1. unsubscribe - to stop to receive commands
1. internal clean up - if something has to be flushed, you are still connected
1. disconnect

Reply:
Issue created - https://github.com/eclipse-kanto/file-upload/issues/31